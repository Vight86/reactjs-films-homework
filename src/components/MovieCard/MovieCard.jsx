import React, { useState } from 'react';
import propTypes from 'prop-types';
import MovieCardGrid from './MovieCardGrid/index';
import MovieCardTable from './MovieCardTable/index';

const MovieCard = ({
  movie, genres, isGrid, movieTrailerKey, updateMovieTrailerKey,
}) => {
  const [isInfoShown, setIsInfoShown] = useState(false);
  const handleShowInfo = () => setIsInfoShown(!isInfoShown);

  const [isModalOpened, setIsModalOpened] = useState(false);

  const handleModalToggle = (movieId, isOpened) => {
    const paddingRight = window.innerWidth - document.documentElement.clientWidth;
    updateMovieTrailerKey(movieId);
    setTimeout(() => { setIsModalOpened(!isOpened); }, 0);
    document.body.style.overflow = !isOpened ? 'hidden' : '';
    document.body.style.paddingRight = !isOpened ? `${paddingRight}px` : '0px';
  };

  const data = {
    movie,
    genres,
    movieTrailerKey,
    updateMovieTrailerKey,
    isInfoShown,
    handleShowInfo,
    isModalOpened,
    handleModalToggle,
  };

  const movieCard = isGrid ? <MovieCardGrid data={data} /> : <MovieCardTable data={data} />;

  return movieCard;
};

MovieCard.propTypes = {
  movie: propTypes.shape({
    adult: propTypes.bool,
    backdropPath: propTypes.string,
    genreIds: propTypes.arrayOf(propTypes.number),
    id: propTypes.number,
    originalLanguage: propTypes.string,
    originalTitle: propTypes.string,
    overview: propTypes.string,
    popularity: propTypes.number,
    posterPath: propTypes.string,
    releaseDate: propTypes.string,
    title: propTypes.string,
    video: propTypes.bool,
    voteAverage: propTypes.number,
    voteCount: propTypes.number,
  }).isRequired,
  genres: propTypes.arrayOf(propTypes.object).isRequired,
  isGrid: propTypes.bool.isRequired,
  movieTrailerKey: propTypes.string,
  updateMovieTrailerKey: propTypes.func,
};

export default MovieCard;
