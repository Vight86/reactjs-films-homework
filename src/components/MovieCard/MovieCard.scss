@import '../../utils/variables';

.viewInfo {
  position: absolute;
  top: 0%;
  left: 0;
  width: 100%;
  height: 100%;
  padding-top: 21%;
  background-color: transparentize($color: $color-dark, $amount: 0.16);
  opacity: 0;
  text-align: center;
  transform: scale(1.25);
  transition: all 0.3s;
}

.movieCard {
  position: relative;
  overflow: hidden;
  padding-bottom: 3.6em;
  background-color: $color-light;
  border-radius: 0.6em;
  box-shadow: 0 10px 40px 0 rgba(0, 0, 0, 0.04);
  font-size: 1rem;

  &:hover {
    .viewInfo {
      opacity: 1;
      transform: scale(1);
    }
  }
}

.poster {
  position: relative;
  overflow: hidden;
  width: 100%;
  margin: 0 0 5.5em 0;
}

.watchNowText {
  display: block;
  margin-bottom: 11%;
  color: $bg-color;
  font-size: 2.8em;
  letter-spacing: 0.04em;
}

.generalInfo {
  display: grid;
  padding: 0 8% 0 8%;
  grid-template-columns: 1fr 20%;
  grid-template-rows: 1fr 1fr;
}

.infoBlock {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  padding-top: 24.3%;
  background:
    linear-gradient(
      transparentize($color: $color-dark, $amount: 0.36),
      transparentize($color: $color-dark, $amount: 0.36)
    ),
    url('./movieStab.jpg');
  background-position: center;
  background-repeat: no-repeat;
  background-size: auto, cover;
  font-size: 1rem;
  opacity: 0;
  transition: all 0.5s;
  visibility: hidden;

  &_active {
    opacity: 1;
    visibility: visible;
  }

  @media screen and (max-width: 980px) {
    font-size: 0.8rem;
  }

  @media screen and (max-width: 320px) {
    font-size: 0.55rem;
  }
}

.infoText {
  overflow: hidden;
  max-height: 8.1em;
  padding: 0 8% 0 8%;
  margin: 9.5% 0 10.7% 0;
  color: $color-light;
  font-size: 2.6em;
  line-height: 1.65;
}
