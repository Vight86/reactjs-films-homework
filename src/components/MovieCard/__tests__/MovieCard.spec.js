import React from 'react';
import Renderer, { act } from 'react-test-renderer';
import MovieCard from '../index';
import image from '../movieStab.jpg';

const data = {
  image,
  title: 'Independance day',
  rating: 3.9,
  genre: ['scifi', 'comedy'],
};

const setUp = (movieData) => {
  const renderer = Renderer.create(<MovieCard movieData={movieData} />);
  return renderer;
};

describe('render MovieCard component', () => {
  it('should be article tag', () => {
    const result = setUp();
    expect(result.toJSON().type).toBe('article');
  });

  it('render component without props', () => {
    const result = setUp();
    expect(result.toJSON()).toMatchSnapshot();
  });

  it('render component with props', () => {
    const result = setUp(data);
    expect(result.toJSON()).toMatchSnapshot();
  });

  it('render View Info button clicks', () => {
    const result = setUp();
    const viewInfoButton = result.root.findByProps({ children: 'View Info' });
    act(() => viewInfoButton.props.onClick());
    expect(result.toJSON()).toMatchSnapshot();
  });

  it('render infoBlock closeButton clicks', () => {
    const result = setUp();
    const viewInfoButton = result.root.findByProps({ children: 'View Info' });
    act(() => viewInfoButton.props.onClick());
    const closeButton = result.root.findByProps({ className: 'closeButton' });
    act(() => closeButton.props.onClick());
    expect(result.toJSON()).toMatchSnapshot();
  });
});
