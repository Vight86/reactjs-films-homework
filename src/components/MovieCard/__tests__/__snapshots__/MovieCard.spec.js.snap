// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render MovieCard component render grid layout correctly 1`] = `
<MovieCardGrid
  data={
    Object {
      "genres": Array [
        Object {
          "id": 1,
          "name": "Action",
        },
        Object {
          "id": 2,
          "name": "SciFi",
        },
      ],
      "handleModalToggle": [Function],
      "handleShowInfo": [Function],
      "isInfoShown": false,
      "isModalOpened": false,
      "movie": Object {
        "adult": false,
        "backdropPath": "/test.jpg",
        "genreIds": Array [
          1,
          2,
        ],
        "id": 1,
        "originalLanguage": "en",
        "originalTitle": "test",
        "overview": "test",
        "popularity": 1,
        "posterPath": "/test.jpg",
        "releaseDate": "1",
        "title": "test",
        "video": false,
        "voteAverage": 1,
        "voteCount": 1,
      },
      "movieTrailerKey": "test",
      "updateMovieTrailerKey": [MockFunction],
    }
  }
/>
`;

exports[`render MovieCard component render handleShowInfo click correctly 1`] = `
<MovieCardTable
  data={
    Object {
      "genres": Array [
        Object {
          "id": 1,
          "name": "Action",
        },
        Object {
          "id": 2,
          "name": "SciFi",
        },
      ],
      "handleModalToggle": [Function],
      "handleShowInfo": [Function],
      "isInfoShown": true,
      "isModalOpened": false,
      "movie": Object {
        "adult": false,
        "backdropPath": "/test.jpg",
        "genreIds": Array [
          1,
          2,
        ],
        "id": 1,
        "originalLanguage": "en",
        "originalTitle": "test",
        "overview": "test",
        "popularity": 1,
        "posterPath": "/test.jpg",
        "releaseDate": "1",
        "title": "test",
        "video": false,
        "voteAverage": 1,
        "voteCount": 1,
      },
      "movieTrailerKey": "test",
      "updateMovieTrailerKey": [MockFunction],
    }
  }
/>
`;

exports[`render MovieCard component render table layout correctly 1`] = `
<MovieCardTable
  data={
    Object {
      "genres": Array [
        Object {
          "id": 1,
          "name": "Action",
        },
        Object {
          "id": 2,
          "name": "SciFi",
        },
      ],
      "handleModalToggle": [Function],
      "handleShowInfo": [Function],
      "isInfoShown": false,
      "isModalOpened": false,
      "movie": Object {
        "adult": false,
        "backdropPath": "/test.jpg",
        "genreIds": Array [
          1,
          2,
        ],
        "id": 1,
        "originalLanguage": "en",
        "originalTitle": "test",
        "overview": "test",
        "popularity": 1,
        "posterPath": "/test.jpg",
        "releaseDate": "1",
        "title": "test",
        "video": false,
        "voteAverage": 1,
        "voteCount": 1,
      },
      "movieTrailerKey": "test",
      "updateMovieTrailerKey": [MockFunction],
    }
  }
/>
`;
